### ISSUE TITLE
Rule suggestion: The '&&' subexpression should be wrapped in parens.

### ISSUE THREAD

**Issue by TimDaub** [eslint/eslint#3270]
Today, a colleague of mine was reviewing my code (we heavily rely on ESLint and it's been awesome for us so far!) and complained about that I'm adding parenthesis to `and` operators that are followed by an `or` operator.
As an example:

```
if(a || b && c)

// should be

if(a || (b && c))
```

Since I totally agreed with him I started to look for rules in ESLint for exactly that. I only found the rule in JSLint (https://github.com/douglascrockford/JSLint/blob/master/jslint.js#L282).

Hence, I wanted to ask about your opinion, if this would be worth adding to ESLint.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by michaelficarra** [comment]
No, this would be adding unnecessary parentheses. We only have a rule that suggests the opposite. I don't think a rule like the one you purpose would be used frequently enough to justify distributing it with eslint.

**Comment by vrde** [comment]
I disagree. They are not **unnecessary**, they help understand the code better, e.g.:

```javascript
// implicit, might be misunderstood:
if (a || b && c) { ... }

// explicit and readable
if (a || (b && c)) { ... }

// explicit and unnecessary
if ((a || (b && c))) { ... }
```

Also, eslint is easily configurable to set a rule off, on as a warning or on with an error, so at the end is up to the user or to the team to decide if they want to enforce the rule :)

Moreover, if we consider stackexchange as a good source of information, [developers agree](http://programmers.stackexchange.com/questions/201175/should-i-use-parentheses-in-logical-statements-even-where-not-necessary) on the use of extra parenthesis to make the code more readable (I suggest to take a look to the post, it is really interesting!).

**Comment by michaelficarra** [comment]
There is nothing to disagree about here. The parentheses are completely, *objectively* unnecessary: they may be removed, and the program will have identical semantics. You are saying they are *useful* for human readers. "Useful" and "necessary" are two distinct concepts. I was not making a judgement about the usefulness of the parentheses in my comment.

I still don't feel this is a common enough desire to justify including a rule like this in eslint. It'd be nice if we had nice, clear guidelines for determining whether a rule qualifies for inclusion in eslint.

**Comment by vrde** [comment]
Using the distinction between *useful* and *necessary*, [many of the rules](http://eslint.org/docs/rules/#best-practices) eslint enforces are *useful* and not *necessary*. Adding extra parenthesis is *useful* for many reasons, that's why I believe it should be added to eslint.

**Comment by michaelficarra** [comment]
Again, I am not disagreeing that some people will believe these unnecessary parentheses are useful. I just don't think *enough* people will believe that. But I have no data that supports that, just my own personal experiences.

**Comment by nzakas** [comment]
Okay, this isn't the way we should be discussing this issue. Rules are not determined by popularity, they are determined by usefulness.

Right now, we have a rule that would flag these parens as unnecessary. So, we can't add a rule that would enforce them otherwise we'd end up with two rules that contradict each other (this is against our policy). The best we could do is have an option that allows this in the current rule, but that is still not the same as enforcing it.

In this case, you may be better off creating a custom rule.

**Comment by ilyavolodin** [comment]
I agree, I think this is a good scenario for a custom rule.

**Comment by gyandeeps** [comment]
Related https://github.com/eslint/eslint/issues/3270

**Comment by nzakas** [comment]
This is #3270 :)

@TimDaub I think the solution here is a custom rule.
